torch/_inductor/runtime/triton_heuristics.py

```diff
@@ -673,7 +682,67 @@ def kernel_call():
 
             return do_bench_using_profiling(kernel_call, warmup=10, rep=40)
 
-        return benchmarker.benchmark_gpu(kernel_call, rep=40)
+        result = benchmarker.benchmark_gpu(kernel_call, rep=40)
+        self.restore_copied_args(cpu_copies)
+        return result
+
+    def copy_args_to_cpu_if_needed(self, *args, **kwargs):
+        copies = {}
+        budget = torch.cuda.max_memory_allocated() - torch.cuda.memory_allocated()
+
+        def maybe_copy(name, arg):
+            if name in self.mutated_arg_names:
+                nonlocal budget
+                assert isinstance(arg, torch.Tensor)
+                assert not arg.is_cpu
+                size = arg.numel() * arg.element_size()
+                if torch.is_floating_point(arg) and size > budget:
+                    cpu_arg = torch.empty_strided(
+                        arg.size(),
+                        arg.stride(),
+                        dtype=arg.dtype,
+                        device="cpu",
+                        pin_memory=True,
+                    )
+                    cpu_arg.copy_(arg, non_blocking=True)
+                    copies[name] = (arg, cpu_arg)
+                else:
+                    budget -= size
+
+        for i, arg in enumerate(args):
+            maybe_copy(self.fn.arg_names[i], arg)
+
+        for name, arg in kwargs.items():
+            maybe_copy(name, arg)
+
+        return copies
+
+    def restore_copied_args(self, cpu_copies):
+        for _, pair in cpu_copies.items():
+            arg, cpu_arg = pair
+            arg.copy_(cpu_arg, non_blocking=True)
+
+    def prepare_args(self, cpu_copies, *args, **kwargs):
+        from ..compile_fx import clone_preserve_strides
+
+        def prepare_arg(name, arg):
+            if name in cpu_copies:
+                assert torch.is_floating_point(arg)
+                arg.uniform_(0, 1)
+                return arg
```

Comment:
What was the perf with actually copying back the values instead of randomizing them ? was there a blocker there ? this would be safer. 

See, we just had this ima internally from bad autotuning assumptions https://fb.workplace.com/groups/1075192433118967/posts/1513763445928528. although note, it actually wasnt related to the values of the tensors, but nonetheless, gives me pause on doing something that is not actually safe. 

There's nothing preventing 
```
indices: float

def kernel(tensor, indices):
     tensor[indices.int()]
```

Maybe skipping tensors with indirect indexing is sufficient.

Reply:
Compile time was quite bad on my devgpu when copying on every iteration, but let's run the full benchmark suite. I'lll be back...